### ISSUE TITLE
com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of org.threeten.bp.OffsetDateTime

### ISSUE THREAD

**Issue by SanalKumarK** [klarna/kco_rest_java#12]
I have upgraded from Klarna version 2.2.0 to 3.0.4 in our project, and followed the examples given in the 3.0.4 archive to create/fetch/update the checkout orders.

POST Request is sent to PATH /checkout/v3/orders and response is received succesfully, but while parsing the JSON response to CheckOutOrder class, I am getting the following errors.

In the following method in CheckoutOrdersAPI class,
public CheckoutOrder create(CheckoutOrder order) throws ApiException, IOException {
        byte[] data = this.objectMapper.writeValueAsBytes(order);
        ApiResponse response = this.post(this.PATH, data);
        response.expectSuccessful().expectStatusCode(Status.CREATED).expectContentType("application/json");
        return **(CheckoutOrder)this.fromJson(response.getBody(), CheckoutOrder.class);**
    }

(CheckoutOrder)this.fromJson(response.getBody(), CheckoutOrder.class) - is throwing following exception.

 com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `org.threeten.bp.OffsetDateTime` (no Creators, like default construct, exist): no String-argument constructor/factory method to deserialize from String value ('2018-12-19T15:33:21Z')
 at [Source: (byte[])"{
  "order_id" : "ec5cb9b1-9ed4-42a2-8978-a807097d5571",
  "status" : "checkout_incomplete",
  "purchase_country" : "se",
  "purchase_currency" : "SEK",
  "locale" : "sv-SE",
  "billing_address" : {
    "country" : "se"
  },
  "customer" : { },
  "shipping_address" : {
    "country" : "se"
  },
  "order_amount" : 14370,
  "order_tax_amount" : 2874,
  "order_lines" : [ {
    "type" : "physical",
    "reference" : "100361005102",
    "name" : "Leggings1_sv (M)",
    "quantity" : 1,
    "quantity_u"[truncated 5354 bytes]; line: 48, column: 18] (through reference chain: com.klarna.rest.api.checkout.model.CheckoutOrder["started_at"])

I updated the jackson.version in our project to 2.9.7 which is used by the Klarna. But still OffsetDateTime parsing throws an exception. 

Any kind of help will be highly appreciated.

**Comment by alexions** [comment]
Hi @SanalKumarK 

Unfortunately, the server response is truncated, so I don't know what exactly you got from the server.
Can you enable the debug mode for SDK and post the full result here?

You can find useful information about how to enable the Debug mode: https://github.com/klarna/kco_rest_java#logging-and-debugging

Will be waiting for your reply.

**Comment by SanalKumarK** [comment]
Hi @alexions ,
I have attached the Response and error message.
From the error message, we can conclude that error is due to jackson deserialization of the field **"started_at"**. Please let us know if we are missing to configure anything for the deserialization of the field "started_at".

[Klarna Response.txt](https://github.com/klarna/kco_rest_java/files/2759332/Klarna.Response.txt)
[Jackson Deserializing error.txt](https://github.com/klarna/kco_rest_java/files/2759333/Jackson.Deserializing.error.txt

**Comment by gfalco77** [comment]
Hi

I have the same issue with OffsetDateTime when I call the PaymentAuthorization API. I'm using v3.0.5 and org.threeten.bp v1.3.8

**Comment by gfalco77** [comment]
I solved it by adding the following libraries together with kco-rest-3.0.5.jar

jackson-annotations-2.9.8.jar
jackson-core-2.9.8.jar
jackson-databind-2.9.8.jar
jackson-datatype-threetenbp-2.8.4.jar
jackson-jaxrs-json-provider-2.9.8.jar
threetenbp-1.3.8.jar

**Comment by alexions** [comment]
@gfalco77 Can you please tell me how did you install the SDK? Did you use Maven/Gradle?

I'm asking because we have exactly the same list of dependencies in the SDK pom.xml: https://github.com/klarna/kco_rest_java/blob/v3.0/pom.xml#L68

So in case of using maven/gradle these dependencies should be auto-imported.

**Comment by SanalKumarK** [comment]
I am using Maven for dependency and build. 
We are using other version of jackson before the klarna upgrade. Since klarna is using the jackson.version 2.9.8, we have upgraded to use the same jackson version. Unfortunately still we are facing the same issue for Klarna upgrade.

**Comment by gfalco77** [comment]
I'm implementing Klarna into SAP Hybris. 
There's a bespoke xml file that looks like a pom. You can run an ant command to download the jars specified there. 
When I debugged, I found that the Klarna response was getting translated by another Jar used in Hybris v.2.9.5. Therefore I thought that if I put in the classpath the referenced jars in the kco-rest-3.0.5.jar pom.xml , it would override the old jars.

**Comment by SanalKumarK** [comment]
Hi @alexions 
As @gfalco77 mentioned, I have added the above dependencies to the POM and tested. But I am still facing the same error. Since the response is parsed by the CheckoutOrdersApi.java create() method, we don't have the control over the json ObjectMapper, to register the threeten module for parsing.
I believe this is something klarna-rest module has to handle.

**Comment by alexions** [comment]
@SanalKumarK 
Talking about the custom ObjectMapper

> we don't have the control over the json ObjectMapper

I created an example to show how you can use your own ObjectMapper: 
https://github.com/klarna/kco_rest_java/blob/v3.0/src/main/java/examples/ObjectMapperExample.java

Using this custom mapper you can do whatewer you want with deserialization and use any custom module.

Let me know if this helped!

**Comment by alexions** [comment]
@SanalKumarK @gfalco77 Hey guys, I forgot to ask very important information. Actually I must start from this question: Which version of JDK do you use?

**Comment by SanalKumarK** [comment]
@alexions 
java version "1.8.0_151"
Java(TM) SE Runtime Environment (build 1.8.0_151-b12)
Java HotSpot(TM) 64-Bit Server VM (build 25.151-b12, mixed mode)

**Comment by buddyp450** [comment]
@alexions 1.8.0_221, what's even stranger is my local development machine is having this problem where another developer's machine is not

**Comment by alexions** [comment]
Hello guys, 

Sorry for the delay in fixing this issue. We are still working on this.
As a quick solution I can offer you to try to do the following:

1) Add a new dependency to your project:
```xml
<dependency>
    <groupId>com.fasterxml.jackson.datatype</groupId>
    <artifactId>jackson-datatype-jsr310</artifactId>
    <version>2.10.1</version>
</dependency>
```

2) Create a new Jackson `ObjectMapper` and add `JavaTimeModule` as a module:
```java
public class MyCustomMapper extends ObjectMapper {
    public MyCustomMapper() {
        this.registerModule(new JavaTimeModule()); // <-- Register a new module
        this.findAndRegisterModules();
        this.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
        this.setSerializationInclusion(JsonInclude.Include.NON_NULL);
        this.setDateFormat(new SimpleDateFormat());
    }
}
```

3) Set your new Custom Mapper to the API entity:
```java
Client client = new Client(username, password, HttpTransport.EU_TEST_BASE_URL);
CheckoutOrdersApi api = client.newCheckoutOrdersApi();
api.setObjectMapper(new MyCustomMapper()); // <-- set a new mapper here
```

After doing this the issue should be fixed.
Please, let me know if this solution helped you, so I will add this to the SDK core.

Best wishes!

**Comment by kasiabanasik** [comment]
@alexions it is working for me. Are you able to tell me when a new version with this fix will be available ?

**Comment by alexions** [comment]
Fixed in v3.2.1
